File: frontend/types/agent.ts
/** 模型内置思考档位：off（默认，最省 token / 最快首 token）/ low / medium / high。 */
export type ThinkingLevel = 'off' | 'low' | 'medium' | 'high';

export interface AgentSkillBinding {
  skill_id: string;
  config: Record<string, unknown>;
  /** 后端在 AgentDetail 里展开的关联 skill（list 接口可能不带） */
  skill?: { slug?: string; name?: string } | null;
}

export interface AgentMCPBinding {
  mcp_server_id: string;
  tool_filter: string[] | null;
}

export type AuxModelRole =
  | 'chat'
  | 'vision'
  | 'image'
  | 'video'
  | 'embedding'
  | 'rerank'
  | 'tts'
  | 'stt'
  | 'custom';

/**
 * Agent / Skill 功能分类（与后端 AgentCategory / SkillCategory 同枚举）。
 * 管理后台按 category 分组渲染。
 */
export type AgentCategory =
  | 'builder'
  | 'installer'
  | 'tester'
  | 'worker.web'
  | 'worker.data'
  | 'worker.io'
  | 'worker.creative'
  | 'utility'
  | 'custom';

/** 分组渲染时使用的中文标签 + 排序顺序。 */
export const AGENT_CATEGORY_LABELS: Record<AgentCategory, string> = {
  builder: 'Builder（编排）',
  installer: 'Installer（技能安装）',
  tester: 'Tester（测试）',
  'worker.web': 'Worker · Web（抓取 / 网络）',
  'worker.data': 'Worker · Data（数据处理）',
  'worker.io': 'Worker · IO（文件 / 通知 / 邮件）',
  'worker.creative': 'Worker · Creative（生成）',
  utility: 'Utility（辅助）',
  custom: 'Custom（未分类）',
};

export const AGENT_CATEGORY_ORDER: AgentCategory[] = [
  'builder',
  'installer',
  'tester',
  'worker.web',
  'worker.data',
  'worker.io',
  'worker.creative',
  'utility',
  'custom',
];

export interface AgentAuxModelBinding {
  model_id: string;
  role: AuxModelRole;
  alias: string | null;
  config: Record<string, unknown>;
}

export interface AgentModelInfo {
  id: string;
  provider_id: string;
  model_id: string;
  display_name: string;
  model_type: 'chat' | 'image' | 'video' | 'embedding' | 'completion';
  context_window: number;
  supports_vision: boolean;
  supports_function_calling: boolean;
}

/** v4: Agent 角色——admin UI 按 kind 分 Super / Worker 两类 */
export type AgentKind = 'super' | 'worker' | 'builder' | 'installer' | 'tester' | 'utility' | null;

export interface AgentPublic {
  id: string;
  name: string;
  description: string;
  category: AgentCategory;
  /** v4 · super / worker / builder / installer / tester / utility（NULL = 未分类老 agent） */
  kind?: AgentKind;
  /** v4 · worker 的 capability slug（如 'xhs_ops'），super 留空 */
  capability?: string | null;
  /** super 的 URL slug（如 Builder Supervisor='builder'）；worker 通常为 null */
  slug?: string | null;
  /** ADR-017 · NULL = 用平台默认模型（运行时按 kind 解析） */
  model_id: string | null;
  soul_md: string;
  protocol_md: string;
  domain_memory_md: string;
  max_iterations: number;
  temperature: number;
  /** 单次 LLM 调用最大输出 token（LiteLLM max_tokens）。默认 5000；命中上限时 ResilientChatLiteLLM 会自动续写（仅纯文本，tool_call 被截会抛错） */
  max_output_tokens: number;
  extra_config: Record<string, unknown>;
  is_enabled: boolean;
  /** ADR-015 · 系统内置对象（不可删除，UI 隐藏删除按钮） */
  is_system?: boolean;
  /** 该 Agent 的 workspace_write 是否视作「交付物」（上 S3、覆盖写、前端显示） */
  produces_deliverable: boolean;
  /** 【旧字段，保留兼容】请改用 thinking_level */
  enable_thinking: boolean;
  /** 模型内置思考档位（off/low/medium/high，默认 off）；_build_llm 按主模型家族映射成各家具体参数 */
  thinking_level: ThinkingLevel;
  created_at: string;
  updated_at: string;
}

export interface AgentDetail extends AgentPublic {
  skill_bindings: AgentSkillBinding[];
  mcp_bindings: AgentMCPBinding[];
  aux_model_bindings: AgentAuxModelBinding[];
  model: AgentModelInfo | null;
}

export interface AgentCreateInput {
  name: string;
  description?: string;
  category?: AgentCategory;
  model_id?: string | null;
  soul_md?: string;
  protocol_md?: string;
  domain_memory_md?: string;
  max_iterations?: number;
  temperature?: number;
  max_output_tokens?: number;
  is_enabled?: boolean;
  produces_deliverable?: boolean;
  enable_thinking?: boolean;
  thinking_level?: ThinkingLevel;
  /** ADR-026 D4 · per-super 配置；含 mission_default_auto_approve 等 key */
  extra_config?: Record<string, unknown>;
}

export type AgentUpdateInput = Partial<AgentCreateInput>;

export interface AgentTestResponse {
  ok: boolean;
  output: string | null;
  tools_loaded: number;
  error: string | null;
}

File: backend/app/schemas/agent.py
"""Agent Pydantic schemas。"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

# 辅助模型角色：chat-model / vision / image-generation / video-generation / embedding / rerank / custom
AuxModelRole = Literal[
    "chat",
    "vision",
    "image",
    "video",
    "embedding",
    "rerank",
    "tts",
    "stt",
    "custom",
]

# Agent / Skill 功能分类，管理后台按此分组渲染。
# - builder: Orchestrator chat 里的"建造者"（meta 层）
# - installer: 安装 ClawHub skill 的执行 Agent（可并行）
# - tester: sandbox 跑 mission_run_test 的 Agent
# - worker.*: 真正执行业务工作流的 Worker（分领域）
# - utility: 通用辅助 Agent（解析、翻译、判断等）
# - custom: 用户自定义未归类
AgentCategory = Literal[
    "builder",
    "installer",
    "tester",
    "worker.web",
    "worker.data",
    "worker.io",
    "worker.creative",
    "worker.imported",  # ADR-019 · 从外部 prompt 库导入的 advisory worker
    "utility",
    "custom",
]


class AgentBase(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    description: str = Field(default="", max_length=512)
    category: AgentCategory = "custom"
    # None = use the platform default model (resolved at runtime by kind). ADR-017.
    model_id: uuid.UUID | None = None
    soul_md: str = ""
    protocol_md: str = ""
    domain_memory_md: str = ""
    # le=80 对齐 SupervisorSpec：Builder 等编排型 super 需更高迭代（reclimit=max_iter*2）。
    max_iterations: int = Field(default=10, ge=1, le=80)
    temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    max_output_tokens: int = Field(
        default=30000,
        ge=256,
        le=64000,
        description=(
            "单次 LLM 调用的最大输出 token（LiteLLM max_tokens）。默认 30000：给交付物 JSON "
            "/ Markdown + LLM 序言 + tool_call 元信息足够余量，彻底规避 length-stop 截断。"
            "命中 length 上限时由 ResilientChatLiteLLM 自动续写（仅纯文本；tool_call 参数被截"
            "会抛错，让 Agent 分块写）。"
        ),
    )
    extra_config: dict = Field(default_factory=dict)
    is_enabled: bool = True
    is_system: bool = Field(
        default=False,
        description="ADR-015 · 平台系统对象（Builder Supervisor / builtin worker）不可删除；前端隐删除钮。",
    )
    produces_deliverable: bool = Field(
        default=False,
        description=(
            "True = 该 Agent 的 workspace_write 被视为交付物：自动上传 S3，"
            "同节点覆盖写（artifacts 数组里同 node 只保留最新一条），前端 Workspace "
            "面板与交付物进度条呈现；False = 过程态，仅保存 branch state。"
        ),
    )
    enable_thinking: bool = Field(
        default=False,
        description="【旧字段，保留兼容】请改用 thinking_level；仅在 thinking_level 缺省时作回退。",
    )
    thinking_level: Literal["off", "low", "medium", "high"] = Field(
        default="off",
        description=(
            "模型内置思考档位。按主模型家族自动映射成各家具体参数：\n"
            "- off    → claude thinking.disabled / gemini thinkingBudget=0(pro 128) / 其它 reasoning_effort=low\n"
            "- low    → gemini thinkingBudget=512  / claude budget_tokens=2000  / reasoning_effort=low\n"
            "- medium → gemini thinkingBudget=2048 / claude budget_tokens=8000  / reasoning_effort=medium\n"
            "- high   → gemini thinkingBudget=8192 / claude budget_tokens=16000 / reasoning_effort=high\n"
            "默认 off（最省 token / 最快首 token）；仍可用 extra_config 精调。"
        ),
    )
    # v3 R24：agent 角色 — worker / super / builder / installer / tester / utility
    kind: str | None = Field(
        default=None,
        description=(
            "v3 角色：worker（被动执行）/ super（主动调度）/ builder / installer / tester / utility。"
            "agent_create 时 None → 由 category 推断（worker.* → 'worker'；builder → 'builder'；"
            "其它 → 'utility'）。super 必须显式传 'super' 触发 R24 强落地默认（max_iterations=40）。"
        ),
    )
    capability: str | None = Field(
        default=None,
        description="worker 能力 slug（如 xhs_ops / data_analytics），供 super invoke_worker 按 capability 选；其它 kind 留空。",
    )


class AgentCreate(AgentBase):
    pass


class AgentUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=512)
    category: AgentCategory | None = None
    model_id: uuid.UUID | None = None
    soul_md: str | None = None
    protocol_md: str | None = None
    domain_memory_md: str | None = None
    max_iterations: int | None = Field(default=None, ge=1, le=80)
    temperature: float | None = Field(default=None, ge=0.0, le=2.0)
    max_output_tokens: int | None = Field(default=None, ge=256, le=64000)
    extra_config: dict | None = None
    is_enabled: bool | None = None
    produces_deliverable: bool | None = None
    enable_thinking: bool | None = None
    thinking_level: Literal["off", "low", "medium", "high"] | None = None
    kind: str | None = None
    capability: str | None = None


class AgentSkillBinding(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    skill_id: uuid.UUID
    config: dict = Field(default_factory=dict)


class AgentMCPBinding(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    mcp_server_id: uuid.UUID
    tool_filter: list[str] | None = None


class AgentAuxModelBinding(BaseModel):
    """Agent 的辅助模型绑定。"""

    model_config = ConfigDict(from_attributes=True)

    model_id: uuid.UUID | None = None
    role: AuxModelRole = "custom"
    alias: str | None = Field(
        default=None, max_length=64, description="Agent 内唯一短名，供工具引用"
    )
    config: dict = Field(default_factory=dict)


class AgentAuxModelBindingInput(BaseModel):
    role: AuxModelRole = "custom"
    alias: str | None = Field(default=None, max_length=64)
    config: dict = Field(default_factory=dict)


class ModelInfo(BaseModel):
    """为前端展示：Detail 中把主模型的关键信息塞进来（provider / model_id / context / vision）。"""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    provider_id: uuid.UUID
    model_id: str
    display_name: str
    model_type: str
    context_window: int = 0
    supports_vision: bool = False
    supports_function_calling: bool = False


class AgentPublic(AgentBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    # super 身份（URL slug + 显示名）；worker/系统对象可能为 None
    slug: str | None = None
    display_name: str | None = None
    created_at: datetime
    updated_at: datetime


class AgentDetail(AgentPublic):
    skill_bindings: list[AgentSkillBinding] = Field(default_factory=list)
    mcp_bindings: list[AgentMCPBinding] = Field(default_factory=list)
    aux_model_bindings: list[AgentAuxModelBinding] = Field(default_factory=list)
    # 主模型扩展信息（UI 展示用）
    model: ModelInfo | None = None


class AgentTestRequest(BaseModel):
    input: str = Field(min_length=1, description="单轮对话输入")


class AgentTestResponse(BaseModel):
    ok: bool
    output: str | None = None
    tools_loaded: int = 0
    error: str | None = None
